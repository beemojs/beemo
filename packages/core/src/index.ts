/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Blueprint, PackageStructure, Path, PortablePath, Predicates } from '@boost/common';

export * from './constants';
export * from './contexts/ConfigContext';
export * from './contexts/Context';
export * from './contexts/DriverContext';
export * from './contexts/ScaffoldContext';
export * from './contexts/ScriptContext';
export * from './Driver';
export * from './Script';
export * from './Tool';
export * from './types';

export { Path };
export type { Blueprint, PackageStructure, PortablePath, Predicates };
