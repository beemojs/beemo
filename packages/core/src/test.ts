import execa from 'execa';
import { ArgList, Arguments, Argv } from '@boost/args';
import { Path } from '@boost/common';
import { mockDebugger } from '@boost/debug/test';
import {
	ConfigContext,
	ConfigFile,
	Context,
	Driver,
	DriverContext,
	DriverContextOptions,
	DriverMetadata,
	ScaffoldContext,
	ScaffoldContextOptions,
	Script,
	ScriptContext,
	ScriptContextOptions,
	Tool,
} from '.';

export { mockDebugger };

const TEST_ROOT = new Path(process.env.BEEMO_TEST_ROOT ?? process.cwd());

export class TestDriver<O extends object = {}> extends Driver<O> {
	override readonly name = 'test-driver';
}

export class TestScript<O extends object = {}> extends Script<O> {
	override readonly name = 'test-script';

	execute(): any {
		return Promise.resolve();
	}
}

export function mockConsole<K extends keyof Console>(name: K): jest.SpyInstance {
	return jest.spyOn(console, name as 'log').mockImplementation(() => {});
}

export function mockToolConfig(): ConfigFile {
	return {
		configure: {
			cleanup: false,
			parallel: true,
		},
		debug: false,
		drivers: [],
		execute: {
			concurrency: 1,
			graph: true,
			output: '',
		},
		module: '@local',
		scripts: [],
		settings: {},
	};
}

export function mockTool(argv: Argv = []): Tool {
	const tool = new Tool({
		argv,
		cwd: TEST_ROOT,
	});

	// @ts-expect-error Allow readonly
	tool.debug = mockDebugger();

	tool.config = mockToolConfig();

	return tool;
}

export function mockDriver<C extends object = {}>(
	name: string,
	tool: Tool | null = null,
	metadata: Partial<DriverMetadata> = {},
): Driver<C> {
	const driver = new TestDriver<C>();

	// @ts-expect-error For testing purposes
	driver.name = name;
	driver.tool = tool ?? mockTool();

	driver.setMetadata({
		bin: name.toLowerCase(),
		configName: `${name}.json`,
		title: name,
		...metadata,
	});

	return driver;
}

export function mockScript(name: string, tool: Tool | null = null): Script<{}> {
	const script = new TestScript<{}>();

	// @ts-expect-error For testing purposes
	script.name = name;
	script.tool = tool ?? mockTool();

	return script;
}

export function stubArgs<T extends object>(options: T, params: ArgList = []): Arguments<T> {
	return {
		command: [],
		errors: [],
		options,
		params,
		rest: [],
		unknown: {},
	};
}

export function stubConfigArgs(): Arguments<{}> {
	return stubArgs({});
}

export function stubDriverArgs(
	fields?: Partial<DriverContextOptions>,
): Arguments<DriverContextOptions> {
	return stubArgs({
		concurrency: 1,
		graph: false,
		workspaces: '',
		...fields,
	});
}

export function stubScaffoldArgs(
	fields?: Partial<ScaffoldContextOptions>,
): Arguments<ScaffoldContextOptions> {
	return stubArgs({
		dry: false,
		...fields,
	});
}

export function stubScriptArgs(
	fields?: Partial<ScriptContextOptions>,
): Arguments<ScriptContextOptions> {
	return stubArgs({
		concurrency: 1,
		graph: false,
		workspaces: '',
		...fields,
	});
}

export function applyContext<T extends Context>(context: T): T {
	context.args = stubArgs({ a: true, foo: 'bar' }, ['baz']);
	context.argv = ['-a', '--foo', 'bar', 'baz'];
	context.cwd = TEST_ROOT;
	context.configModuleRoot = TEST_ROOT;
	context.workspaceRoot = TEST_ROOT;
	context.workspaces = [];

	return context;
}

export function stubConfigContext(): ConfigContext {
	return applyContext(new ConfigContext(stubArgs({})));
}

export function stubDriverContext(driver?: Driver): DriverContext {
	return applyContext(new DriverContext(stubDriverArgs(), driver ?? new TestDriver()));
}

export function stubScriptContext(script?: Script): ScriptContext {
	const context = applyContext(new ScriptContext(stubScriptArgs(), 'script'));

	if (script) {
		context.setScript(script);
	}

	return context;
}

export function stubScaffoldContext(
	generator: string = 'generator',
	action: string = 'action',
	name: string = '',
): ScaffoldContext {
	return applyContext(new ScaffoldContext(stubScaffoldArgs(), generator, action, name));
}

export function prependRoot(part: string): Path {
	return TEST_ROOT.append(part);
}

export function getRoot(): Path {
	return TEST_ROOT;
}

export function stubExecResult(fields?: Partial<execa.ExecaReturnValue>): execa.ExecaReturnValue {
	return {
		all: '',
		command: '',
		escapedCommand: '',
		exitCode: 0,
		failed: false,
		isCanceled: false,
		killed: false,
		signal: undefined,
		stderr: '',
		stdout: '',
		timedOut: false,
		...fields,
	};
}
