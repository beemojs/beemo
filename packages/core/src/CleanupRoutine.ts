import { Routine } from '@boost/core';
import chalk from 'chalk';
import fs from 'fs-extra';
import DriverContext from './contexts/DriverContext';
import { BeemoTool } from './types';

export default class CleanupRoutine extends Routine<DriverContext, BeemoTool> {
  bootstrap() {
    this.task(this.tool.msg('app:configCleanup'), this.deleteConfigFiles);
  }

  /**
   * Delete all temporary config files.
   */
  async deleteConfigFiles(context: DriverContext): Promise<boolean[]> {
    return Promise.all(
      context.configPaths.map(config => {
        this.debug('Deleting config file %s', chalk.cyan(config.path));

        this.tool
          .getPlugin('driver', config.driver)
          .onDeleteConfigFile.emit([context, config.path]);

        return fs.remove(config.path).then(() => true);
      }),
    );
  }
}
