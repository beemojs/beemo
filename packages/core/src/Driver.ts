import execa from 'execa';
import mergeWith from 'lodash/mergeWith';
import { Blueprint, isObject, optimal, Path, Predicates, predicates } from '@boost/common';
import { ConcurrentEvent, Event } from '@boost/event';
import { Plugin } from '@boost/plugin';
import {
  STRATEGY_COPY,
  STRATEGY_CREATE,
  STRATEGY_NATIVE,
  STRATEGY_NONE,
  STRATEGY_REFERENCE,
} from './constants';
import ConfigContext from './contexts/ConfigContext';
import DriverContext from './contexts/DriverContext';
import isClassInstance from './helpers/isClassInstance';
import Tool from './Tool';
import {
  Argv,
  BeemoTool,
  Driverable,
  DriverCommandOptions,
  DriverMetadata,
  DriverOptions,
  DriverOutput,
  DriverStrategy,
  Execution,
} from './types';

export default abstract class Driver<
    Config extends object = {},
    Options extends DriverOptions = DriverOptions
  >
  extends Plugin<BeemoTool, Options>
  implements Driverable {
  command: DriverCommandOptions = {};

  // Set after instantiation
  config!: Config;

  // Set after instantiation
  metadata!: DriverMetadata;

  // Set within a life-cycle
  tool!: Tool;

  output: DriverOutput = {
    stderr: '',
    stdout: '',
  };

  readonly onLoadProviderConfig = new Event<[ConfigContext, Path, Config]>('load-provider-config');

  readonly onLoadConsumerConfig = new Event<[ConfigContext, Config]>('load-consumer-config');

  readonly onMergeConfig = new Event<[ConfigContext, Config]>('merge-config');

  readonly onCreateConfigFile = new Event<[ConfigContext, Path, Config]>('create-config-file');

  readonly onCopyConfigFile = new Event<[ConfigContext, Path, Config]>('copy-config-file');

  readonly onReferenceConfigFile = new Event<[ConfigContext, Path, Config]>(
    'reference-config-file',
  );

  readonly onDeleteConfigFile = new Event<[ConfigContext, Path]>('delete-config-file');

  readonly onBeforeExecute = new ConcurrentEvent<[DriverContext, Argv]>('before-execute');

  readonly onAfterExecute = new ConcurrentEvent<[DriverContext, unknown]>('after-execute');

  readonly onFailedExecute = new ConcurrentEvent<[DriverContext, Error]>('failed-execute');

  static validate(driver: Driver) {
    const name = (isClassInstance(driver) && driver.constructor.name) || 'Driver';

    if (!isObject(driver.options)) {
      throw new Error(`\`${name}\` requires an options object.`);
    }
  }

  blueprint({ array, object, string }: Predicates): Blueprint<DriverOptions> {
    return {
      args: array(string()),
      dependencies: array(string()),
      env: object(string()),
      strategy: string(STRATEGY_NATIVE).oneOf<DriverStrategy>([
        STRATEGY_NATIVE,
        STRATEGY_CREATE,
        STRATEGY_REFERENCE,
        STRATEGY_COPY,
        STRATEGY_NONE,
      ]),
    };
  }

  bootstrap() {}

  startup(tool: Tool) {
    this.tool = tool;
    this.bootstrap();
  }

  /**
   * Special case for merging arrays.
   */
  doMerge(prevValue: unknown, nextValue: unknown): unknown {
    if (Array.isArray(prevValue) && Array.isArray(nextValue)) {
      return Array.from(new Set(prevValue.concat(nextValue)));
    }

    return undefined;
  }

  /**
   * Extract the error message when the driver fails to execute.
   */
  extractErrorMessage(error: { message: string }): string {
    return error.message.split('\n', 1)[0] || '';
  }

  /**
   * Format the configuration file before it's written.
   */
  formatConfig(data: Config): string {
    const content = JSON.stringify(data, null, 2);

    if (this.metadata.configName.endsWith('.js')) {
      return `module.exports = ${content};`;
    }

    return content;
  }

  /**
   * Return the module name without the Beemo namespace.
   */
  getName(): string {
    return this.name.split('-').pop()!;
  }

  /**
   * Return a list of user defined arguments.
   */
  getArgs(): Argv {
    return Array.isArray(this.options.args) ? this.options.args : [];
  }

  /**
   * Return a list of dependent drivers.
   */
  getDependencies(): string[] {
    const dependencies = Array.isArray(this.options.dependencies) ? this.options.dependencies : [];

    return [
      // Always required; configured by the driver
      ...this.metadata.dependencies,
      // Custom; configured by the consumer
      ...dependencies,
    ];
  }

  /**
   * Return a list of supported CLI options.
   */
  getSupportedOptions(): string[] {
    return [];
  }

  /**
   * Extract the current version of the installed driver via its binary.
   */
  getVersion(): string {
    const { bin, versionOption } = this.metadata;
    const version = (execa.sync(bin, [versionOption])?.stdout || '').trim();
    const match = version.match(/(\d+)\.(\d+)\.(\d+)/u);

    return match ? match[0] : '0.0.0';
  }

  /**
   * Merge multiple configuration objects.
   */
  mergeConfig(prev: Config, next: Config): Config {
    return mergeWith(prev, next, this.doMerge);
  }

  /**
   * Handle command failures according to this driver.
   */
  processFailure(error: Execution) {
    const { stderr = '', stdout = '' } = error;
    const out = (stderr || stdout).trim();

    if (out) {
      this.setOutput('stderr', out);
    }
  }

  /**
   * Handle successful commands according to this driver.
   */
  processSuccess(response: Execution) {
    const { stderr = '', stdout = '' } = response;

    this.setOutput('stderr', stderr.trim());
    this.setOutput('stdout', stdout.trim());
  }

  /**
   * Setup additional command options.
   */
  setCommandOptions(options: DriverCommandOptions): this {
    const blueprint: Blueprint<DriverCommandOptions> = {};
    const { shape, string } = predicates;

    Object.keys(options).forEach((key) => {
      blueprint[key] = shape({
        description: string().notEmpty().required(),
        type: string().oneOf<'boolean' | 'number' | 'string'>(['string', 'number', 'boolean']),
      });
    });

    this.command = optimal(options, blueprint, {
      name: this.constructor.name,
      unknown: true,
    });

    return this;
  }

  /**
   * Set metadata about the binary/executable in which this driver wraps.
   */
  setMetadata(metadata: Partial<DriverMetadata>): this {
    const { array, bool, string } = predicates;

    this.metadata = optimal(
      metadata,
      {
        bin: string()
          .match(/^[a-z]{1}[a-zA-Z0-9-]+$/u)
          .required(),
        configName: string().required(),
        configOption: string('--config'),
        configStrategy: string(STRATEGY_CREATE).oneOf([
          STRATEGY_CREATE,
          STRATEGY_REFERENCE,
          STRATEGY_COPY,
        ]),
        dependencies: array(string()),
        description: string(),
        filterOptions: bool(true),
        helpOption: string('--help'),
        title: string().required(),
        useConfigOption: bool(),
        versionOption: string('--version'),
        watchOptions: array(string()),
        workspaceStrategy: string(STRATEGY_REFERENCE).oneOf([STRATEGY_REFERENCE, STRATEGY_COPY]),
      },
      {
        name: this.constructor.name,
      },
    );

    return this;
  }

  /**
   * Store the raw output of the driver's execution.
   */
  setOutput(type: keyof DriverOutput, value: string): this {
    this.output[type] = value.trim();

    return this;
  }
}
