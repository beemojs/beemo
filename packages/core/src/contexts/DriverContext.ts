import ConfigContext from './ConfigContext';
import Driver from '../Driver';
import { Arguments, Argv, StdioType } from '../types';

export interface DriverContextOptions {
  concurrency: number;
  graph: boolean;
  stdio: StdioType;
  workspaces: string;
}

export default class DriverContext extends ConfigContext<DriverContextOptions> {
  // Name defined on the plugin (kebab case)
  driverName: string = '';

  // Nested list of argv for each parallel execution
  parallelArgv: Argv[] = [];

  // The primary driver that initiated the pipeline
  primaryDriver: Driver;

  constructor(args: Arguments<DriverContextOptions>, driver: Driver, parallelArgv: Argv[] = []) {
    super(args);

    this.driverName = driver.name;
    this.parallelArgv = parallelArgv;
    this.primaryDriver = driver;

    // Add primary driver to driver list
    this.drivers.add(driver);
  }

  /**
   * Add a parallel command with additional argv.
   */
  addParallelCommand(argv: Argv): this {
    this.parallelArgv.push(argv);

    return this;
  }
}
