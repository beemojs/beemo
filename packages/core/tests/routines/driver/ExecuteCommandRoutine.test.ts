import execa from 'execa';
import fs from 'fs-extra';
import { ExitError, Path } from '@boost/common';
import { normalizeSeparators } from '@boost/common/test';
import { color } from '@boost/internal';
import { DriverContext } from '../../../src/contexts/DriverContext';
import { Driver } from '../../../src/Driver';
import { ExecuteCommandRoutine } from '../../../src/routines/driver/ExecuteCommandRoutine';
import {
	getRoot,
	mockDebugger,
	mockDriver,
	mockTool,
	prependRoot,
	stubDriverContext,
} from '../../../src/test';
import { Tool } from '../../../src/Tool';

describe('ExecuteCommandRoutine', () => {
	let routine: ExecuteCommandRoutine;
	let context: DriverContext;
	let driver: Driver;
	let tool: Tool;

	beforeEach(() => {
		tool = mockTool();

		driver = mockDriver('babel', tool);
		driver.configure({
			args: ['--qux'],
			env: { DEV: 'true' },
		});

		context = stubDriverContext(driver);

		routine = new ExecuteCommandRoutine('babel', 'Run babel', {
			argv: ['-a', '--foo', 'bar', 'baz'],
			tool,
		});
		// @ts-expect-error Overwrite readonly
		routine.debug = mockDebugger();
	});

	describe('constructor()', () => {
		it('errors if `forceConfigOption` is not a boolean', () => {
			expect(() => {
				routine = new ExecuteCommandRoutine('test', 'test', {
					// @ts-expect-error Invalid type
					forceConfigOption: 'foo',
				});
			}).toThrowErrorMatchingSnapshot();
		});

		it('errors if `packageRoot` is not a string', () => {
			expect(() => {
				routine = new ExecuteCommandRoutine('test', 'test', {
					// @ts-expect-error Invalid type
					packageRoot: 123,
				});
			}).toThrowErrorMatchingSnapshot();
		});
	});

	describe('captureOutput()', () => {
		let logSpy: jest.SpyInstance;
		let errorSpy: jest.SpyInstance;
		let stream: execa.ExecaChildProcess;

		class MockStream {
			handler?: (chunk: Buffer) => void;

			pipe() {
				return this;
			}

			on(key: string, handler: (chunk: Buffer) => void) {
				this.handler = handler;

				return this;
			}

			emit() {
				if (this.handler) {
					this.handler(Buffer.from('buffered', 'utf8'));
				}
			}
		}

		beforeEach(() => {
			tool.outStream = { write() {} };
			tool.errStream = { write() {} };

			stream = {
				// @ts-expect-error Invalid type
				stdout: new MockStream(),
				// @ts-expect-error Invalid type
				stderr: new MockStream(),
			};

			logSpy = jest.spyOn(tool.outStream, 'write');
			errorSpy = jest.spyOn(tool.errStream, 'write');
		});

		afterEach(() => {
			logSpy.mockRestore();
			errorSpy.mockRestore();
		});

		describe('watch', () => {
			it('enables if args option matches watch option', () => {
				driver.metadata.watchOptions = ['--watch'];
				context.args.unknown.watch = 'true';

				expect(routine.captureOutput(context, stream)).toBe('watch');
			});

			it('enables if args option matches short watch option', () => {
				driver.metadata.watchOptions = ['-w'];
				context.args.unknown.w = 'true';

				expect(routine.captureOutput(context, stream)).toBe('watch');
			});

			it('enables if positional args includes watch option', () => {
				driver.metadata.watchOptions = ['watch'];
				context.args.params = ['watch'];

				expect(routine.captureOutput(context, stream)).toBe('watch');
			});

			it('disables if args option doesnt match watch option', () => {
				driver.metadata.watchOptions = ['--watch'];

				expect(routine.captureOutput(context, stream)).toBe('buffer');
			});

			it('disables if positional args doesnt include watch option', () => {
				driver.metadata.watchOptions = ['--watch'];
				context.args.params = ['--notWatch'];

				expect(routine.captureOutput(context, stream)).toBe('buffer');
			});

			it('disables if args option is falsy', () => {
				driver.metadata.watchOptions = ['--watch'];

				expect(routine.captureOutput(context, stream)).toBe('buffer');
			});

			it('pipes a batch stream when enabled', () => {
				const outSpy = jest.spyOn(stream.stdout!, 'pipe');
				const errSpy = jest.spyOn(stream.stderr!, 'pipe');

				driver.metadata.watchOptions = ['--watch'];
				context.args.unknown.watch = 'true';

				routine.captureOutput(context, stream);

				expect(outSpy).toHaveBeenCalled();
				expect(errSpy).toHaveBeenCalled();
			});

			it('registers a data handler when using watch', () => {
				const outSpy = jest.spyOn(stream.stdout!, 'on');
				const errSpy = jest.spyOn(stream.stderr!, 'on');

				driver.metadata.watchOptions = ['--watch'];
				context.args.unknown.watch = 'true';

				routine.captureOutput(context, stream);

				expect(outSpy).toHaveBeenCalledWith('data', expect.anything());
				expect(errSpy).toHaveBeenCalledWith('data', expect.anything());
			});

			it('writes chunk to stdout stream', () => {
				driver.metadata.watchOptions = ['--watch'];
				context.args.unknown.watch = 'true';

				routine.captureOutput(context, stream);

				stream.stdout!.emit('data');

				expect(logSpy).toHaveBeenCalledWith('buffered');
			});
		});

		['stream', 'pipe'].forEach((strategy) => {
			describe(`${strategy}`, () => {
				beforeEach(() => {
					driver.configure({
						outputStrategy: strategy as 'stream',
					});
				});

				it('enables if option is set', () => {
					expect(routine.captureOutput(context, stream)).toBe(strategy);
				});

				it('doesnt pipe a batch stream', () => {
					const outSpy = jest.spyOn(stream.stdout!, 'pipe');
					const errSpy = jest.spyOn(stream.stderr!, 'pipe');

					routine.captureOutput(context, stream);

					expect(outSpy).not.toHaveBeenCalled();
					expect(errSpy).not.toHaveBeenCalled();
				});

				it('registers a data handler', () => {
					const outSpy = jest.spyOn(stream.stdout!, 'on');
					const errSpy = jest.spyOn(stream.stderr!, 'on');

					routine.captureOutput(context, stream);

					expect(outSpy).toHaveBeenCalledWith('data', expect.anything());
					expect(errSpy).toHaveBeenCalledWith('data', expect.anything());
				});

				it('writes chunk to stdout stream', () => {
					routine.captureOutput(context, stream);

					stream.stdout!.emit('data');

					expect(logSpy).toHaveBeenCalledWith('buffered');
				});
			});
		});

		describe('buffer', () => {
			beforeEach(() => {
				driver.configure({
					outputStrategy: 'buffer',
				});
			});

			it('defaults to buffer if no option or not watching', () => {
				expect(routine.captureOutput(context, stream)).toBe('buffer');
			});
		});
	});

	describe('execute()', () => {
		beforeEach(() => {
			driver.metadata.filterOptions = true;
		});

		it('executes pipeline in order', async () => {
			routine.options.argv.push('--out-dir', './lib');

			const argSpy = jest.spyOn(routine, 'gatherArgs');
			const globSpy = jest.spyOn(routine, 'expandGlobPatterns');
			const filterSpy = jest.spyOn(routine, 'filterUnknownOptions');
			const optSpy = jest.spyOn(routine, 'includeConfigOption');
			const runSpy = jest.spyOn(routine, 'runCommandWithArgs');

			const response = await routine.execute(context);

			expect(argSpy).toHaveBeenCalledWith(context, undefined, expect.anything());
			expect(globSpy).toHaveBeenCalledWith(
				context,
				['--qux', '-a', '--foo', 'bar', 'baz', '--out-dir', './lib'],
				expect.anything(),
			);
			expect(filterSpy).toHaveBeenCalledWith(
				context,
				['--qux', '-a', '--foo', 'bar', 'baz', '--out-dir', './lib'],
				expect.anything(),
			);
			expect(optSpy).not.toHaveBeenCalled();
			expect(runSpy).toHaveBeenCalledWith(
				context,
				['baz', '--out-dir', './lib'],
				expect.anything(),
			);
			expect(response).toEqual({ stdout: '' });
		});

		it('includes config option if `useConfigOption` is true', async () => {
			driver.metadata.useConfigOption = true;

			context.configPaths.push({
				driver: 'babel',
				path: prependRoot(driver.metadata.configName),
			});

			const optSpy = jest.spyOn(routine, 'includeConfigOption');
			const runSpy = jest.spyOn(routine, 'runCommandWithArgs');

			await routine.execute(context);

			expect(optSpy).toHaveBeenCalledWith(context, ['baz'], expect.anything());
			expect(runSpy).toHaveBeenCalledWith(
				context,
				['baz', '--config', prependRoot(driver.metadata.configName).path()],
				expect.anything(),
			);
		});

		it('doesnt filter unknown if `filterOptions` is false', async () => {
			driver.metadata.filterOptions = false;

			const filterSpy = jest.spyOn(routine, 'filterUnknownOptions');

			await routine.execute(context);

			expect(filterSpy).not.toHaveBeenCalled();
		});

		it('calls `copyConfigToWorkspace` when driver is workspaces enabled', async () => {
			driver.metadata.workspaceStrategy = 'copy';
			routine.configure({
				packageRoot: '/some/root',
			});

			const copySpy = jest.spyOn(routine, 'copyConfigToWorkspacePackage');

			await routine.execute(context);

			expect(copySpy).toHaveBeenCalledWith(context, ['baz'], expect.anything());
		});

		it('doesnt call `copyConfigToWorkspace` when driver is not workspaces enabled', async () => {
			routine.configure({
				packageRoot: '/some/root',
			});

			const copySpy = jest.spyOn(routine, 'copyConfigToWorkspacePackage');

			await routine.execute(context);

			expect(copySpy).not.toHaveBeenCalled();
		});

		it('doesnt call `copyConfigToWorkspace` when no workspace root', async () => {
			driver.metadata.workspaceStrategy = 'copy';

			const copySpy = jest.spyOn(routine, 'copyConfigToWorkspacePackage');

			await routine.execute(context);

			expect(copySpy).not.toHaveBeenCalled();
		});

		it('doesnt call `expandGlobPatterns` when expandGlobs is `false`', async () => {
			driver.configure({ expandGlobs: false });

			const expandGlobsSpy = jest.spyOn(routine, 'expandGlobPatterns');

			await routine.execute(context);

			expect(expandGlobsSpy).not.toHaveBeenCalled();
		});
	});

	describe('copyConfigToWorkspacePackage()', () => {
		it('copies each config into workspace root', async () => {
			const copySpy = jest.spyOn(fs, 'copyFileSync').mockImplementation(() => true);

			routine.configure({
				packageRoot: '/some/root',
			});

			context.configPaths = [
				{ driver: 'babel', path: new Path('.babelrc') },
				{ driver: 'jest', path: new Path('jest.json') },
			];

			const args = await routine.copyConfigToWorkspacePackage(context, ['foo', '--bar']);

			expect(args).toEqual(['foo', '--bar']);
			expect(copySpy).toHaveBeenCalledWith('.babelrc', normalizeSeparators('/some/root/.babelrc'));
			expect(copySpy).toHaveBeenCalledWith(
				'jest.json',
				normalizeSeparators('/some/root/jest.json'),
			);

			copySpy.mockRestore();
		});
	});

	describe('expandGlobPatterns()', () => {
		it('passes through if no globs', async () => {
			const args = await routine.expandGlobPatterns(context, ['--foo', 'bar', '-z']);

			expect(args).toEqual(['--foo', 'bar', '-z']);
		});

		it('converts globs to paths', async () => {
			const args = await routine.expandGlobPatterns(context, ['--foo', '../tests/*', 'bar']);

			// Make testing easier
			args.sort();

			expect(args).toEqual([
				'--foo',
				normalizeSeparators('../tests/__fixtures__'),
				normalizeSeparators('../tests/configs'),
				normalizeSeparators('../tests/setup.ts'),
				'bar',
			]);
		});

		it('handles missing paths', async () => {
			const args = await routine.expandGlobPatterns(context, ['../some-fake-path/*.js']);

			expect(args).toEqual([]);
		});
	});

	describe('extractNativeOptions()', () => {
		it('extracts all types of options', async () => {
			const options = await routine.extractNativeOptions(context);

			expect(options).toEqual(
				expect.objectContaining({
					'-f': true, // Short
					'-M': true, // Uppercase short
					'--filename': true, // Long
					'--no-highlight-code': true, // With dashes
				}),
			);
		});

		it('supports multiple options within `helpOption`', async () => {
			driver.metadata.helpOption = '--help --all';

			const spy = jest.spyOn(routine, 'executeCommand');

			await routine.extractNativeOptions(context);

			expect(spy).toHaveBeenCalledWith('babel', ['--help', '--all'], {
				env: { DEV: 'true' },
				preferLocal: true,
			});
		});

		it('supports uppercased options', async () => {
			driver.getSupportedOptions = () => ['-u', '--all', '--changedFiles', '--watch-only'];

			context.primaryDriver = driver;

			const options = await routine.extractNativeOptions(context);

			expect(options).toEqual(
				expect.objectContaining({
					'-u': true, // Short
					'--all': true, // Long
					'--changedFiles': true, // Camel case
					'--watch-only': true, // Dashed
				}),
			);
		});
	});

	describe('filterUnknownOptions()', () => {
		it('returns supported options', async () => {
			const args = await routine.filterUnknownOptions(context, [
				'./src',
				'-o',
				'./lib',
				'--out-dir',
				'./dist',
				'--minified',
			]);

			expect(args).toEqual(['./src', '-o', './lib', '--out-dir', './dist', '--minified']);
		});

		it('filters unsupported options', async () => {
			const args = await routine.filterUnknownOptions(context, [
				'./src',
				'--foo',
				'-o',
				'./lib',
				'--out-dir',
				'./dist',
				'-X',
				'--minified',
				'--bar',
			]);

			expect(args).toEqual(['./src', '-o', './lib', '--out-dir', './dist', '--minified']);
			expect(routine.debug).toHaveBeenCalledWith(
				'Filtered args: %s',
				color.mute('--foo, -X, --bar'),
			);
		});

		it('skips unsupported option setters', async () => {
			const args = await routine.filterUnknownOptions(context, [
				'--foo',
				'123',
				'--bar=456',
				'-w',
				'-c',
				'789',
				'-c=666',
			]);

			expect(args).toEqual(['-w']);
			expect(routine.debug).toHaveBeenCalledWith(
				'Filtered args: %s',
				color.mute('--foo, 123, --bar=456, -c, 789, -c=666'),
			);
		});
	});

	describe('gatherArgs()', () => {
		it('merges driver and command line args', async () => {
			const args = await routine.gatherArgs(context);

			expect(args).toEqual(['--qux', '-a', '--foo', 'bar', 'baz']);
		});

		it('rebuilds context args object', async () => {
			expect(context.args).toEqual(
				expect.objectContaining({
					params: ['baz'],
					options: {
						a: true,
						foo: 'bar',
					},
				}),
			);

			await routine.gatherArgs(context);

			expect(context.args).toEqual(
				expect.objectContaining({
					params: ['baz'],
					options: {
						a: true,
						foo: 'bar',
						qux: true,
					},
				}),
			);
		});
	});

	describe('includeConfigOption()', () => {
		it('does nothing if a config path doesnt match', async () => {
			const args = await routine.includeConfigOption(context, ['--foo']);

			expect(args).toEqual(['--foo']);
		});

		it('appends config path for a match', async () => {
			context.configPaths.push({
				driver: 'babel',
				path: prependRoot(driver.metadata.configName),
			});

			const args = await routine.includeConfigOption(context, ['--foo']);

			expect(args).toEqual(['--foo', '--config', prependRoot(driver.metadata.configName).path()]);
		});
	});

	describe('runCommandWithArgs()', () => {
		const task = expect.anything(); // new Task<DriverContext>('Task', () => {});

		beforeEach(() => {
			jest
				.spyOn(routine, 'executeCommand')
				.mockImplementation(() => Promise.resolve({ success: true } as any));
			jest.spyOn(driver, 'processSuccess').mockImplementation();
			jest.spyOn(driver, 'processFailure').mockImplementation();
		});

		it('executes command with correct args', async () => {
			await routine.runCommandWithArgs(context, ['--wtf'], task);

			expect(routine.executeCommand).toHaveBeenCalledWith('babel', ['--wtf'], {
				cwd: getRoot().path(),
				env: { DEV: 'true' },
				preferLocal: true,
				workUnit: task,
				wrap: expect.any(Function),
			});
		});

		it('handles success using driver', async () => {
			const response = await routine.runCommandWithArgs(context, ['--wtf'], task);

			expect(response).toEqual({ success: true });
			expect(driver.processSuccess).toHaveBeenCalledWith({ success: true });
		});

		it('handles failure using driver', async () => {
			(routine.executeCommand as jest.Mock).mockImplementation(() =>
				Promise.reject(new Error('Oops')),
			);

			try {
				await routine.runCommandWithArgs(context, ['--wtf'], task);
			} catch (error) {
				expect(driver.processFailure).toHaveBeenCalledWith(error);
			}
		});

		it('persists exit code when a failure', async () => {
			(routine.executeCommand as jest.Mock).mockImplementation(() => {
				const error = new Error('Oops');
				// @ts-expect-error Field doesnt exist on errors
				error.exitCode = 3;

				return Promise.reject(error);
			});

			try {
				await routine.runCommandWithArgs(context, ['--wtf'], task);
			} catch (error) {
				expect(error).toBeInstanceOf(ExitError);
				expect((error as ExitError).code).toBe(3);
			}
		});

		it('handles out of memory failures', async () => {
			(routine.executeCommand as jest.Mock).mockImplementation(() =>
				// eslint-disable-next-line prefer-promise-reject-errors
				Promise.reject({ exitCode: null, message: '', signal: 'SIGKILL' }),
			);

			try {
				await routine.runCommandWithArgs(context, ['--wtf'], task);
			} catch (error) {
				expect(error).toEqual(new ExitError('Out of memory!', 1));
			}
		});

		it('doesnt handle failure using driver if MaxBufferError occured', async () => {
			const error = new Error('Oops');
			error.name = 'MaxBufferError';

			(routine.executeCommand as jest.Mock).mockImplementation(() => Promise.reject(error));

			try {
				await routine.runCommandWithArgs(context, ['--wtf'], task);
			} catch {
				expect(driver.processFailure).not.toHaveBeenCalled();
			}
		});

		it('emits `onBeforeExecute` event', async () => {
			const spy = jest.fn();

			driver.onBeforeExecute.listen(spy);

			await routine.runCommandWithArgs(context, ['--wtf'], task);

			expect(spy).toHaveBeenCalledWith(context, ['--wtf']);
		});

		it('emits `onAfterExecute` event on success', async () => {
			const spy = jest.fn();

			driver.onAfterExecute.listen(spy);

			await routine.runCommandWithArgs(context, ['--wtf'], task);

			expect(spy).toHaveBeenCalledWith(context, { success: true });
		});

		it('emits `onFailedExecute` event on failure', async () => {
			const spy = jest.fn();

			driver.onFailedExecute.listen(spy);

			(routine.executeCommand as jest.Mock).mockImplementation(() =>
				Promise.reject(new Error('Oops')),
			);

			try {
				await routine.runCommandWithArgs(context, ['--wtf'], task);
			} catch (error) {
				expect(spy).toHaveBeenCalledWith(context, error);
			}
		});
	});
});
