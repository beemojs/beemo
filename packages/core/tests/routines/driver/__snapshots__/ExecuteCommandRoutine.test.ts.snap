// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExecuteCommandRoutine constructor() errors if \`forceConfigOption\` is not a boolean 1`] = `
"The following validations have failed for \`ExecuteCommandRoutine\`:
  - Invalid field \\"forceConfigOption\\". Must be a boolean, received string.
  - Invalid field \\"tool\\". Field is required and must be defined."
`;

exports[`ExecuteCommandRoutine constructor() errors if \`packageRoot\` is not a string 1`] = `
"The following validations have failed for \`ExecuteCommandRoutine\`:
  - Invalid field \\"packageRoot\\". Must be a string, received number.
  - Invalid field \\"tool\\". Field is required and must be defined."
`;
